### PR TITLE
fix(compiler): support async globalScripts functions

### LINKED ISSUE #3392: bug: `globalScript` not supporting `async`
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/stencil/blob/main/.github/CONTRIBUTING.md).
- [X] I agree to follow the [Code of Conduct](https://github.com/ionic-team/stencil/blob/main/CODE_OF_CONDUCT.md).
- [X] I have searched for [existing issues](https://github.com/ionic-team/stencil/issues) that already report this problem, without success.

### Stencil Version

2.15.2

### Current Behavior

according to the [`globalScript` documentation](https://stenciljs.com/docs/config#globalscript) the exported method supports `async`, however when testing this the bootstrapping is not_ async.

With a little bit debugging I can see that the generated bootstrap methods won't `await` my code.

```
const globalScripts = () => {
  appComponentsGlobalScript();
  initialize();
};
```

```
patchBrowser().then(options => {
  globalScripts();
  return bootstrapLazy(JSON.parse('lotsOfJsonHere'), options);
});
```

### Expected Behavior

That my `async` code is properly `await`ed before `bootstrapLazy` is executed

### Steps to Reproduce

just export an `async` method in `globalScripts` file and put something in `componentDidLoad` on a component - check order of execution.

I've done this in the [reprocuction URL](https://codesandbox.io/s/stencil-globalscript-bug-r87gn2) with four `console.log` statements.

The expected order should be:

```
global:start 
global:end
component:willLoad 
component:didLoad 
```

The current order:

```
global:start 
component:willLoad 
component:didLoad 
global:end 
```

### Code Reproduction URL

https://codesandbox.io/s/stencil-globalscript-bug-r87gn2

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by christian-bromann**
## What is the current behavior?
Scripts defined in the `globalScript` Stencil configuration won't delay the execution of the initialization process. However this is what [the Stencil docs](https://stenciljs.com/docs/config#globalscript) suggest

fixes #3392
STENCIL-467

## What is the new behavior?
With this patch Stencil will start recognize an asynchronous function and will wait for the async promise to resolve before continuing the initialisation.

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

This could be a surprise for people that do async operations in these scripts but where not concerned about having their components being rendered before.

## Testing

I added an e2e test but you can verify this patch by:

- creating a new stencil project
- add a `globalScript` that does async stuff (see test example)
- observe that components are immediately rendered vs. stale rendering when compiling with this branch

## Other information

While looking for a fix for this issue I started to doubt the usefulness of this feature. I probably don't have enough context for how global scripts in Stencil are used but I believe there is are barely any reasons why you would stale the rendering of your application like that.

As an alternative I suggest to update our documentation to not promise async behavior. I assume this won't be a suprise for anyone since it didn't work before.

### CODE REVIEW

**Review by alicewriteswrongs — APPROVED** [review]
lgtm!

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]
Two additional asks:
1. Can we update the PR + commit message to use semantic commits so this gets registered properly by the changelog
2. Can we add a warning that this is a blocking call in https://stenciljs.com/docs/config#globalscript? In the repro case of the original issue they use 5 seconds as the delay, which although arbitrary, is quite a long time and made me concerned as first this wasn't working 😅

File: src/compiler/bundle/app-data-plugin.ts [hunk 1]
```diff
@@ -183,9 +183,11 @@ const appendGlobalScripts = (globalScripts: GlobalScript[], s: MagicString) => {
     });
 
     s.append(`export const globalScripts = () => {\n`);
+    s.append(`  return Promise.all([`);
     globalScripts.forEach((globalScript) => {
-      s.append(`  ${globalScript.defaultName}();\n`);
+      s.append(`    ${globalScript.defaultName}(),\n`);
     });
+    s.append(`  ]);\n`);
```
[the remark this answers is not in the document]
Comment (christian-bromann) [reply src/compiler/bundle/app-data-plugin.ts:190]: I missed a `\n` here, will add it.

File: test/karma/test-app/util.ts [hunk 2]
```diff
@@ -17,7 +17,7 @@ export declare namespace SomeTypes {
 /**
  * Utilities for creating a test bed to execute HTML rendering tests against
  */
-type DomTestUtilities = {
+export type DomTestUtilities = {
```
[the remark this answers is not in the document]
Comment (christian-bromann) [reply test/karma/test-app/util.ts:20]: Seemed like my IDE was confused here about this. Safe to revert.

**Comment by christian-bromann** [comment]
Thanks for the review, I addressed the comments. 

> 1. Can we update the PR + commit message to use semantic commits so this gets registered properly by the changelog

Done 👍 

> 2\. Can we add a warning that this is a blocking call

Raised a PR with a proposal: https://github.com/ionic-team/stencil-site/pull/1350

File: src/testing/puppeteer/puppeteer.d.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+/**
```
Comment (rwaskiewicz) [inline src/testing/puppeteer/puppeteer.d.ts:1]: Non-blocking question I missed the first time around - do we need this file?

**Review by rwaskiewicz — APPROVED** [review]

File: src/testing/puppeteer/puppeteer.d.ts (hunk 3 above)
Comment (christian-bromann) [reply src/testing/puppeteer/puppeteer.d.ts]: I have no idea how this got in here 😆 removing it.

**Comment by mikaelkaron** [comment]
Side note (since I’m the original reporter of #3392).

A use-case for supporting `async` here would be supporting loading configuration needed _before_ an application is started that might affect what to load down the line (i18n packages, dark/light mode only components, etc..).

You can argue that it’s possible to structure the app to display a loader while information like this is fetched, but you can _also_ do that by adding the loader to the `html` itself and hide/remove it once the app is loaded.

In summary, you can probably work around this, but it’s really nice to have this supported by the compiler.

**Comment by christian-bromann** [comment]
Thanks for the feedback @mikaelkaron 🙏
